pip/pip-331.md

```diff
@@ -0,0 +1,129 @@
+# PIP-331: WASM Function API
+
+# Background knowledge
+
+WASM(WebAssembly) bytecode is designed to be encoded in a size- and load-time-efficient binary format. WASM aims to leverage the common hardware features available on various platforms to execute in browsers at machine code speed.
+
+WASI(WebAssembly System Interface) provide a portable interface for applications that run within a constrained sandbox environment, which allows WASM to run in non browser environments such as Linux. It's portable and secure.
+
+# Motivation
+
+The server and client sides of the Pulsar function use protobuf for decoupling. In principle, the language supported by protobuf can be supported by the pulsar function, now Pulsar provided the java, python and golang function client, but there are still many languages that are not supported.
+
+Before all language adaptations are completed (and it's almost entirely certain to be impossible), users cannot write pulsar function in their familiar languages.
+
+# Goals
+
+## In Scope
+
+Other languages, as long as their code can be compiled into WASM bytecode (such as Rust/golang/C++), users can use these languages to write pulsar function.
+
+## Out of Scope
+
+All existing abilities of the Java pulsar function client are not reimplemented, the WASM Pulsar functions is under the Java Pulsar functions.
+
+Due to the strict requirements of WASM on parameter types and for simplicity reasons, types other than `java.lang.Long` are not used as parameters or return value.
+
+# High Level Design
+
+```mermaid 
+flowchart LR;
+
+    subgraph develop
+        direction TB
+        SourceCode ==> |"CompileToWASM"| WasmFile ==> |"RenameFile"| MoveToTheResourceDirectory ==> UnitTest
+    end
+
+    subgraph runtime
+        direction TB
+        PulsarFunctionJava ==> |"LoadFromResource"| TheWasmFile ==> |"Invoke"| TheSourceCode
+    end
+    
+    develop --> runtime
+```
+
+# Detailed Design
+
+## Design & Implementation Details
+
+1. add `WasmLoader` to load WASM file and provide the WASM function to java, also provide the java function to WASM if we need.
+
+2. add `AbstractWasmFunction` and `AbstractWasmWindowFunction` as the core interface of the WASM function api.
+
+```java
+public abstract class AbstractWasmFunction<X, T> extends WasmLoader implements Function<X, T> {
+
+    private static final String PROCESS_METHOD_NAME = "process";
+
+    protected static final String INITIALIZE_METHOD_NAME = "initialize";
+
+    protected static final String CLOSE_METHOD_NAME = "close";
+
+    protected static final Map<Long, Argument<?>> ARGUMENTS = new ConcurrentHashMap<>();
+
+    @Override
+    public T process(X input, Context context) {
+        return super.getWasmExtern(PROCESS_METHOD_NAME)
+                .map(process -> {
+                    Long argumentId = callWASI(input, context, process);
```

Comment:
What exactly happens here? Does it load a virtual machine first that can execute WASM code?

Reply:
The WASM bytecode has been loaded, here we directly obtain the abstract WASM function for Java to call.

We load the WASM bytecode in `WasmLoader#WasmLoader`.